lxd/metrics/api_rates.go

```diff
@@ -0,0 +1,102 @@
+package metrics
+
+import (
+	"net/url"
+	"sync/atomic"
+
+	"github.com/canonical/lxd/shared/api"
+	"github.com/canonical/lxd/shared/entity"
+)
+
+// RequestResult represents a completed request status category.
+type RequestResult string
+
+const (
+	errorServer RequestResult = "error_server"
+	errorClient RequestResult = "error_client"
+	success     RequestResult = "succeeded"
+)
+
+// requestResultsTypes contains all the possible request result categories.
+var requestResultsTypes = []RequestResult{
+	errorServer,
+	errorClient,
+	success,
+}
+
+// GetRequestResultsTypes returns all possible request result types.
+func GetRequestResultsTypes() []RequestResult {
+	return requestResultsTypes
+}
+
+// ResolveResponseStatus categorizes a status code from a http response as serverError, clientError or success.
+func ResolveResponseStatus(statusCode int) RequestResult {
+	// 4* codes are considered client errors on HTTP.
+	if statusCode >= 400 && statusCode < 500 {
+		return errorClient
+	}
+
+	// 2* codes are considered successes on HTTP, 300 are for forwarded requests
+	// that are included as successes for the purpose of the metrics.
+	if statusCode >= 200 && statusCode < 400 {
+		return success
+	}
+
+	// Any other status code is considered a server error.
+	return errorServer
+}
+
+// ResolveOperationStatus categorizes a status code from an operation as serverError, clientError or success.
+func ResolveOperationStatus(statusCode api.StatusCode) RequestResult {
+	// If the operation was not completed as expected an neither cancelled by an user, it is considered a failure.
+	if statusCode != api.Success && statusCode != api.Cancelled {
+		return errorServer
+	}
+
+	// Otherwise we consider it a success.
+	return success
+}
+
+type completedMetricsLabeling struct {
+	entityType entity.Type
+	result     RequestResult
+}
+
+var ongoingRequests map[entity.Type]*atomic.Int64
+var completedRequests map[completedMetricsLabeling]*atomic.Int64
+
+// InitAPIMetrics initializes maps with initial values for the API rates metrics.
+func InitAPIMetrics() {
+	relevantEntityTypes := entity.APIMetricsEntityTypes()
+	ongoingRequests = make(map[entity.Type]*atomic.Int64, len(relevantEntityTypes))
+	completedRequests = make(map[completedMetricsLabeling]*atomic.Int64, len(relevantEntityTypes)*len(requestResultsTypes))
+
+	for _, entityType := range relevantEntityTypes {
+		ongoingRequests[entityType] = new(atomic.Int64)
+		for _, result := range requestResultsTypes {
+			completedRequests[completedMetricsLabeling{entityType: entityType, result: result}] = new(atomic.Int64)
```

Comment:
Can we avoid pre-allocating and pre-zeroing and instead generate missing ones during the metrics generation output to save memory?

Reply:
What is the memory consumption  for these 2 new fully initialised maps?

Reply:
I decided for this approach because then the increment and decrement of the values becomes as efficient and simple as possible (as far as I see at least). And the memory consuption should be small, measuring `runtime.MemStats` before and after populating the map returns a difference of less than 500 bytes.

If you would like me to try to measure this some other way please let me know.

Reply:
This measurement was done after apllying your suggestion on changing `RequestResult` to an int.